sycl/doc/extensions/proposed/sycl_ext_oneapi_address_cast.asciidoc

```diff
@@ -0,0 +1,204 @@
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+The `sycl::address_space_cast` function in SYCL 2020 does two things: 1) checks
+whether a given raw pointer can be cast to a specific address space; and 2)
+performs the casting operation. In cases where the developer is attempting to
+assert that a raw pointer points to an object in a specific address space, the
+checks from 1) are not required and may have undesirable performance impact.
+
+This extension separates `sycl::address_space_cast` into two functions:
+
+- `static_address_cast`, which casts with no run-time checks.
+- `dynamic_address_cast`, which casts with run-time checks.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ADDRESS_CAST` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Address space cast functions
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+static_address_cast(ElementType* ptr);
+
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+dynamic_address_cast(ElementType* ptr);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+[source,c++]
+----
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+static_address_cast(ElementType* ptr);
+----
+_Preconditions_: `ptr` points to an object allocated in the address space
+designated by `Space`.
+
+_Returns_: A `multi_ptr` with the specified address space and decoration that
+points to the same object as `ptr`.
+
+[NOTE]
+====
+Implementations using address space deduction may choose to issue a diagnostic
```

Comment:
Presumably, it would be OK for any implementation to issue a diagnostic if it can prove that `ptr` does not point to the expected address space.  The spec says that the behavior is UB in this case, so it seems like issuing a diagnostic would be allowed.

Reply:
You're right.  I've modified this to cover all implementations in a462056f2866.